// @sudoku/stores/strategy_display.js
import { writable } from 'svelte/store';
import {manager} from "@sudoku/strategy/strategy_manager"

function createStrategyDisplay() {
    const strategy_display = writable({});

    return {
        subscribe: strategy_display.subscribe,


        update(board) {
            strategy_display.clear;
            strategy_display.update($strategy_display => {
                $strategy_display.clear;
                let dict = manager.getDictFromMatrix(board, 1); //这里的board要求是二维数组
                for (const key in dict) {
                    $strategy_display[key] = dict[key][0];
                }

                return $strategy_display;
            });
        },

        clear() {
            strategy_display.clear;
        }
    }
}

export const strategy_display = createStrategyDisplay();

export function updateUserGridFromCoordinates() {
    strategy_display.subscribe((coordinates) => {


        coordinates.forEach(({ x, y, value }) => {
            if ( value.length === 1) {
                // 当 value 长度为 1 时，调用 userGrid.set
                userGrid.set({ x, y }, value[0]);
            }
            else if (value.length > 1) {
                // 当 value 包含多个元素时，逐个调用 candidates.add
                value.forEach(singleValue => {
                    candidates.add({ x, y }, singleValue); // 对每个值调用 candidates.add
                });
            }
        });
    });
}
