// @sudoku/strategy/strategy.js

// 抽象策略类
export class Strategy {
    constructor(name) {
        if (this.constructor === Strategy) {
            throw new Error('Abstract classes can\'t be instantiated.');
        }
        this.name = name;
    }

    get_candidate(board) {
        throw new Error('Abstract method must be implemented in derived classes.');
    }
}

// 策略管理器类
export class StrategyManager {
    constructor() {
        this.strategies = [];
        this.candidateLists = [];
    }

    // 添加策略
    addStrategy(strategy) {
        if (!(strategy instanceof Strategy)) {
            throw new Error('Only instances of Strategy can be added.');
        }
        this.strategies.push(strategy);
    }

    // 根据策略的名称进行删除
    removeStrategy(strategy_name) {
        for(let index=0; index<=this.strategies.length; index++) {
            if(this.strategies[index].name == strategy_name){
                delete this.strategies[index--];
            }
        }
    }

    // 根据board获取不同策略下的candidate
    executeStrategies(board) {
        this.candidateLists = [];
        for (const strategy of this.strategies) {
            const candidates = strategy.get_candidate(board);
            this.candidateLists.push(candidates);
        }
        return this.getIntersectionCandidates();
    }

    // 对candidateLists取交集
    getIntersectionCandidates() {
        if (this.candidateLists.length === 0) {
            return {};
        }
        
        let newIntersection = this.candidateLists[0];

        for (let i = 1; i < this.candidateLists.length; i++) {
            for (let row=0; row<9; row++) {
                for (let col=0; col<9; col++) {
                    newIntersection[row][col] = newIntersection[row][col].filter(num => this.candidateLists[i][row][col].includes(num))
                }
            }
        }

        return newIntersection;
    }
}

export const manager = new StrategyManager();
